hydra-node/src/Hydra/Chain/Direct/Tx.hs

```diff
@@ -8,60 +10,140 @@ module Hydra.Chain.Direct.Tx where
-    , txProtocolParams = BuildTxWith Nothing
-    , txWithdrawals = TxWithdrawalsNone
-    , txCertificates = TxCertificatesNone
-    , txUpdateProposal = TxUpdateProposalNone
-    , txMintValue = TxMintNone
-    , txScriptValidity = BuildTxWith TxScriptValidityNone
+initTx :: HeadParameters -> TxIn StandardCrypto -> ValidatedTx (AlonzoEra StandardCrypto)
+initTx HeadParameters{contestationPeriod, parties} txIn =
+  mkUnsignedTx body dats
+ where
+  body =
+    TxBody
+      (Set.singleton txIn) -- inputs
+      mempty -- collateral
+      (StrictSeq.singleton headOut) -- outputs
+      mempty -- txcerts
+      (Wdrl mempty) -- txwdrls
+      (Coin 0) -- txfee
+      (ValidityInterval SNothing SNothing) -- txvldt
+      SNothing -- txUpdates
+      mempty -- reqSignerHashes
+      mempty -- mint
+      SNothing -- scriptIntegrityHash
+      SNothing -- adHash
+      SNothing -- txnetworkid
+      --
+  dats = TxDats $ Map.singleton headDatumHash headDatum
+
+  headOut = TxOut headAddress headValue (SJust headDatumHash)
+
+  -- TODO(SN): The main Hydra Head script address. Will be parameterized by the
+  -- thread token eventually. For now, this is just some arbitrary address, as
+  -- it is also later quite arbitrary/different per Head.
+  headAddress :: Addr StandardCrypto
+  headAddress =
+    Addr
+      network
+      (ScriptHashObj $ hashScript @(AlonzoEra StandardCrypto) headScript)
+      -- REVIEW(SN): stake head funds?
+      StakeRefNull
+
+  -- REVIEW(SN): how much to store here / minUtxoValue / depending on assets?
+  headValue = inject (Coin 0)
+
+  headDatumHash = hashData @(AlonzoEra StandardCrypto) headDatum
+
+  headDatum =
+    Data . toData . toBuiltinData $
+      Initial
+        (contestationPeriodFromDiffTime contestationPeriod)
+        (map (partyFromVerKey . vkey) parties)
+
+  headScript = PlutusScript "some invalid plutus script"
+
+-- * Observe Hydra Head transactions
+
+observeTx :: ValidatedTx (AlonzoEra StandardCrypto) -> Maybe (OnChainTx tx)
+observeTx tx =
+  observeInitTx tx
+    <|> observeCommitTx tx
+
+observeInitTx :: ValidatedTx (AlonzoEra StandardCrypto) -> Maybe (OnChainTx tx)
+observeInitTx ValidatedTx{wits} = do
+  fromData plutusData >>= \case
+    Initial cp ps ->
+      pure $ OnInitTx (contestationPeriodToDiffTime cp) (map convertParty ps)
+    _ -> Nothing
+ where
+  -- TODO(SN): temporary hack
+  [(_, Data plutusData)] = Map.toList . unTxDats $ txdats wits
+
+  convertParty = anonymousParty . partyToVerKey
+
+observeCommitTx :: ValidatedTx (AlonzoEra StandardCrypto) -> Maybe (OnChainTx tx)
+observeCommitTx _ = Nothing
+--
+
+-- * Helpers
+
+mkUnsignedTx ::
+  TxBody (AlonzoEra StandardCrypto) ->
+  TxDats (AlonzoEra StandardCrypto) ->
+  ValidatedTx (AlonzoEra StandardCrypto)
+mkUnsignedTx body datums =
+  ValidatedTx
+    { body
+    , wits =
+        TxWitness
+          mempty -- txwitsVKey
+          mempty -- txwitsBoot
+          mempty --txscripts
+          datums -- txdats
```

Comment:
question: Do I understand correctly this field should actually be used for _inputs_' datums, not for outputs' ones?

Reply:
I remember some discussions about having optional inclusion of UTXO's datums within the producing transaction and not the consuming one. I don't know technically where those would be placed though, perhaps in `auxData`?

Reply:
> question: Do I understand correctly this field should actually be used for inputs' datums, not for outputs' ones?

Yes and no. You can also use it for "output datums", but then, their hash need to be referenced in the `scriptIntegrityHashes` in the TxBody (without what, nodes may drop them from the witness list without making the transaction invalid since they would be non-required witnesses).

Reply:
I am not computing any `scriptIntegrityHashes` and would need a bit more explanation / help on this I suppose :)